### PR TITLE
add support for moondream vision language model

### PR DESCRIPTION
**Pull request by vikhyat**
This required making the following changes to the CLIP model:

1. Support for patch embedding bias.
2. Make class embedding and pre-layernorm optional.
3. Add support for post-layernorm.

I verified that the LLaVA model still works as expected after this change.

### CODE REVIEW

File: examples/llava/clip.cpp [hunk 1]
```diff
@@ -1148,12 +1170,39 @@ struct clip_ctx * clip_model_load(const char * fname, const int verbosity = 1) {
 
         }
 
+        try {
+            vision_model.class_embedding  = get_tensor(new_clip->ctx_data, TN_CLASS_EMBD);
+            new_clip->has_class_embedding = true;
+        } catch(const std::exception& e) {
```
Comment (ggerganov) [inline examples/llava/clip.cpp:1176]:
```suggestion
        } catch (const std::exception& e) {
```

**Review by ggerganov — APPROVED** [review]

**Comment by CoderCowMoo** [comment]
Would this be a general support for all SigLIP based encoders or just for moondream2?

**Comment by s-smits** [comment]
Unfortunately, I can't convert the model to GGUF with gguf-my-repo; would it be better to build the source myself (with Cmake) for a higher chance of success?
Screenshot 2024-10-05 at 14 01 36

**Comment by Mihailoff** [comment]
I wonder will the new multimodal implementation https://github.com/ggml-org/llama.cpp/blob/master/docs/multimodal.md work for `moondream`?

**Pull request merged**